import-meta-path-helpers.md

```diff
@@ -0,0 +1,76 @@
+- `__filename` is a string like `/Users/Geoffrey/my-project/src/index.js`
+- `__dirname` is a string like `/Users/Geoffrey/my-project/src`
+
+Since `__filename` and `__dirname` have been around [since Node.js 0.1.27](https://nodejs.org/api/modules.html#__dirname), released in 2010, they are very well known to Node.js developers and very broadly referenced by `npm` packages. It is common to find documentation suggesting passing `__dirname` into a function for serving the static assets of a folder, for example, or passing `__filename` into a function used for logging.
+
+In Node.js, developers working with such path-based APIs need to jump through a few hoops. Node.js provides a helper `fileURLToPath` that developers can import from `node:url`, to enable code like `serveStaticAssets(fileURLToPath(dirname(import.meta.url)))`. A common complaint of Node.js developers is that this is too verbose, and a way to achieve the same with simpler code and without needing to import utility helpers would be preferable.
+
+## Proposal
+
+For runtimes that support filesystem-based modules, we propose adding two new optional properties to `import.meta`:
+
+### `import.meta.filename`
+
+A string containing the full absolute filesystem path to the current module, like CommonJS `__filename`. This is the same as `fileURLToPath(import.meta.url)` in Node.js. An example value would be `/Users/Geoffrey/my-project/src/index.js`.
+
+The value will be undefined for modules that are not loaded from the filesystem, such as modules whose `import.meta.url` begins with `http` or `data`.
+
+### `import.meta.dirname`
+
+A string containing the full absolute filesystem path of the folder containing the current module, like CommonJS `__dirname`. This is the same as `dirname(fileURLToPath(import.meta.url))` in Node.js. An example value would be `/Users/Geoffrey/my-project/src`.
+
+The value should not have a trailing slash. This is consistent with CommonJS `__dirname`, so that it can be used as a drop-in replacement.
+
+The value will be undefined for modules that are not loaded from the filesystem, such as modules whose `import.meta.url` begins with `http` or `data`.
+
+## Notes
+
+These properties would be optional, even if a runtime supports filesystem-based modules. This allows for runtimes where modules are _primarily_ run from HTTP / HTTPS URLs to avoid creating a portability hazard where these properties exist while developers are working locally but are undefined when the code is deployed.
+
+This portability hazard is a primary objection to this proposal. We feel that on balance, the convenience of these properties outweighs the potential hazard; and that a different portability hazard exists without these helpers, in that developers will write code such as `fileURLToPath(import.meta.url)` that works locally but not over CDN similarly to how `import.meta.filename` would work locally but not over CDN.
```

Comment:
I really wish we have clarification and standards about Windows path normalization and consider forward slash possibility as standard once we have a chance with new spec.

An example URL would be `file:///C:/path/` and normalized today as `C:\path` with Node.js utils.

Modern Windows systems well support forward slashes ([some history](https://learn.microsoft.com/en-us/archive/blogs/larryosterman/why-is-the-dos-path-character)). Forward slashes make life much better in the Javascript ecosystem where as backslash is a escape character. [upath](https://github.com/anodynos/upath) and [pathe](https://github.com/unjs/pathe) are already being used by >18M users weekly.

(sorry for rambling in the comments, just wanted to trigger a discussion from somewhere)

Reply:
Related: https://github.com/denoland/deno/issues/957, https://github.com/nodejs/node/pull/25324.

Reply:
I had the same thought too at first, but now I’m not so sure. I think the existing behavior is because Node’s `path` module assembles the parts of the path, then does essentially `parts.join(path.sep)`, and the path separator for Windows is a backslash. It could very well be that this is 13-year-old behavior for old versions of Windows that aren’t relevant anymore, I don’t know. I also wonder if it would be bad to be inconsistent within the same runtime: currently in Node, `__filename` and `fileURLToPath(import.meta.url)` return backslash-delimited paths in Windows, so having a forward-slash path in `import.meta.filename` might be surprising. As far as the spec is concerned, maybe this is best left as an implementation detail for hosts to determine; other runtimes without path-based APIs that use backslashes might be freer to standardize on forward slashes everywhere. Though I’m open to defining the slashes as needing to be forward slashes if those who know this topic better than me can confirm that it’s safe and wise.

Reply:
Thanks for the references. I think it might be sensible for Node.js which is inheriting legacy reasons for preserving backslashes (while I hope that would be something they might consider for the new `import.meta` API for the sake of portability...). 

Since this is a proposal, would it makes sense to at least mention the possibility of a "recommendation" to prefer forward slashes and increasing web/url compatibility? Especially with new runtimes adopting `import.meta` props, they can do it in a non breaking sense.